File: android/app/src/main/java/com/openburnbar/data/media/FileSealAEAD.kt
package com.openburnbar.data.media

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

object FileSealAEAD {
    const val CHUNK_PLAINTEXT_BYTES = 32 * 1024 * 1024
    const val NONCE_SIZE = 12
    const val TAG_SIZE = 16
    const val MAX_PLAINTEXT_BYTES = 10L * 1024L * 1024L * 1024L

    private const val CONTENT_KEY_BYTES = 32
    private const val TAG_SIZE_BITS = TAG_SIZE * Byte.SIZE_BITS

    /** AAD suffix after the attachment id: chunkIndex + totalChunks + plaintextSize, all big-endian longs. */
    private const val AAD_SUFFIX_BYTES = 3 * Long.SIZE_BYTES

    data class Header(
        val attachmentId: String,
        val totalChunks: Int,
        val plaintextSize: Long,
        val contentBlake3: String,
    )

    fun mintContentKey(): ByteArray = ByteArray(CONTENT_KEY_BYTES).also { SecureRandom().nextBytes(it) }

    fun mintNonce(): ByteArray = ByteArray(NONCE_SIZE).also { SecureRandom().nextBytes(it) }

    fun aad(header: Header, chunkIndex: Long): ByteArray {
        val id = header.attachmentId.toByteArray(Charsets.UTF_8)
        val buf = ByteBuffer.allocate(id.size + AAD_SUFFIX_BYTES).order(ByteOrder.BIG_ENDIAN)
        buf.put(id)
        buf.putLong(chunkIndex)
        buf.putLong(header.totalChunks.toLong())
        buf.putLong(header.plaintextSize)
        return buf.array()
    }

    fun sealChunk(plaintext: ByteArray, contentKey: ByteArray, header: Header, chunkIndex: Long, nonce: ByteArray): Pair<ByteArray, ByteArray> {
        require(nonce.size == NONCE_SIZE)
        require(contentKey.size == CONTENT_KEY_BYTES)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(contentKey, "AES"), GCMParameterSpec(TAG_SIZE_BITS, nonce))
        cipher.updateAAD(aad(header, chunkIndex))
        val combined = cipher.doFinal(plaintext)
        val ciphertext = combined.copyOfRange(0, combined.size - TAG_SIZE)
        val tag = combined.copyOfRange(combined.size - TAG_SIZE, combined.size)
        return ciphertext to tag
    }

    fun openChunk(ciphertext: ByteArray, tag: ByteArray, contentKey: ByteArray, header: Header, chunkIndex: Long, nonce: ByteArray): ByteArray {
        require(nonce.size == NONCE_SIZE)
        require(contentKey.size == CONTENT_KEY_BYTES)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(contentKey, "AES"), GCMParameterSpec(TAG_SIZE_BITS, nonce))
        cipher.updateAAD(aad(header, chunkIndex))
        return cipher.doFinal(ciphertext + tag)
    }

    fun sealFile(source: File, destination: File, contentKey: ByteArray, header: Header) {
        require(header.plaintextSize <= MAX_PLAINTEXT_BYTES)
        FileInputStream(source).use { input ->
            FileOutputStream(destination).use { output ->
                val buf = ByteArray(CHUNK_PLAINTEXT_BYTES)
                var index = 0L
                while (true) {
                    val read = input.read(buf)
                    if (read <= 0) break
                    val chunk = buf.copyOf(read)
                    val nonce = mintNonce()
                    val sealed = sealChunk(chunk, contentKey, header, index, nonce)
                    output.write(nonce)
                    output.write(sealed.first)
                    output.write(sealed.second)
                    index += 1
                }
                require(index.toInt() == header.totalChunks)
            }
        }
    }

    fun openFile(source: File, destination: File, contentKey: ByteArray, header: Header) {
        FileInputStream(source).use { input ->
            FileOutputStream(destination).use { output ->
                var remaining = header.plaintextSize
                for (index in 0 until header.totalChunks) {
                    val plaintextLen = minOf(CHUNK_PLAINTEXT_BYTES.toLong(), remaining).toInt()
                    val wireLen = NONCE_SIZE + plaintextLen + TAG_SIZE
                    val wire = ByteArray(wireLen)
                    val read = input.read(wire)
                    require(read == wireLen)
                    val nonce = wire.copyOfRange(0, NONCE_SIZE)
                    val ciphertext = wire.copyOfRange(NONCE_SIZE, wireLen - TAG_SIZE)
                    val tag = wire.copyOfRange(wireLen - TAG_SIZE, wireLen)
                    val plain = openChunk(ciphertext, tag, contentKey, header, index.toLong(), nonce)
                    output.write(plain)
                    remaining -= plain.size
                }
                require(remaining == 0L)
            }
        }
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarMedia/FileSealAEAD.swift
import Foundation
import OpenBurnBarKernel

/// OBFS1 — chunked AES-GCM file seal. Random 12-byte IVs are stored with each
/// chunk. Never HKDF-derive the nonce; a retry mints a new IV.
public enum FileSealAEAD {
    public static let chunkPlaintextBytes = 32 * 1024 * 1024
    public static let nonceSize = 12
    public static let tagSize = 16
    public static let maxPlaintextBytes: Int64 = 2 * 1024 * 1024 * 1024

    public enum Error: Swift.Error, Equatable {
        case invalidNonce
        case invalidKey
        case truncatedChunk
        case chunkCountMismatch
        case plaintextTooLarge
    }

    public struct Header: Equatable, Sendable {
        public var attachmentId: String
        public var totalChunks: Int
        public var plaintextSize: Int64
        public var contentBlake3: String

        public init(attachmentId: String, totalChunks: Int, plaintextSize: Int64, contentBlake3: String) {
            self.attachmentId = attachmentId
            self.totalChunks = totalChunks
            self.plaintextSize = plaintextSize
            self.contentBlake3 = contentBlake3
        }
    }

    public static func mintContentKey() throws -> Data {
        try PlatformCrypto.secureRandomBytes(count: 32)
    }

    public static func mintNonce() throws -> Data {
        try PlatformCrypto.secureRandomBytes(count: nonceSize)
    }

    public static func aad(header: Header, chunkIndex: UInt64) -> Data {
        var data = Data(header.attachmentId.utf8)
        func appendBE(_ value: UInt64) {
            var be = value.bigEndian
            withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
        }
        appendBE(chunkIndex)
        appendBE(UInt64(header.totalChunks))
        appendBE(UInt64(bitPattern: header.plaintextSize))
        return data
    }

    public static func sealChunk(
        plaintext: Data,
        contentKey: Data,
        header: Header,
        chunkIndex: UInt64,
        nonce: Data
    ) throws -> (ciphertext: Data, tag: Data) {
        guard nonce.count == nonceSize else { throw Error.invalidNonce }
        guard contentKey.count == 32 else { throw Error.invalidKey }
        let sealed = try PlatformCrypto.sealAESGCMDetached(
            plaintext: plaintext,
            keyData: contentKey,
            nonce: nonce,
            authenticating: aad(header: header, chunkIndex: chunkIndex)
        )
        return (sealed.ciphertext, sealed.tag)
    }

    public static func openChunk(
        ciphertext: Data,
        tag: Data,
        contentKey: Data,
        header: Header,
        chunkIndex: UInt64,
        nonce: Data
    ) throws -> Data {
        guard nonce.count == nonceSize else { throw Error.invalidNonce }
        guard contentKey.count == 32 else { throw Error.invalidKey }
        return try PlatformCrypto.openAESGCMDetached(
            nonce: nonce,
            ciphertext: ciphertext,
            tag: tag,
            keyData: contentKey,
            authenticating: aad(header: header, chunkIndex: chunkIndex)
        )
    }

    /// Stream-seal `source` into `destination` without loading the whole file.
    public static func sealFile(
        from source: URL,
        to destination: URL,
        contentKey: Data,
        header: Header
    ) throws {
        guard header.plaintextSize <= maxPlaintextBytes else { throw Error.plaintextTooLarge }
        let inHandle = try FileHandle(forReadingFrom: source)
        defer { try? inHandle.close() }
        FileManager.default.createFile(atPath: destination.path, contents: nil)
        let outHandle = try FileHandle(forWritingTo: destination)
        defer { try? outHandle.close() }
        var index: UInt64 = 0
        while true {
            let chunk = inHandle.readData(ofLength: chunkPlaintextBytes)
            if chunk.isEmpty { break }
            let nonce = try mintNonce()
            let sealed = try sealChunk(
                plaintext: chunk,
                contentKey: contentKey,
                header: header,
                chunkIndex: index,
                nonce: nonce
            )
            var wire = nonce
            wire.append(sealed.ciphertext)
            wire.append(sealed.tag)
            outHandle.write(wire)
            index += 1
        }
        guard Int(index) == header.totalChunks else { throw Error.chunkCountMismatch }
    }

    /// Stream-open `source` into `destination` without loading the whole ciphertext.
    public static func openFile(
        from source: URL,
        to destination: URL,
        contentKey: Data,
        header: Header
    ) throws {
        let inHandle = try FileHandle(forReadingFrom: source)
        defer { try? inHandle.close() }
        FileManager.default.createFile(atPath: destination.path, contents: nil)
        let outHandle = try FileHandle(forWritingTo: destination)
        defer { try? outHandle.close() }
        var remaining = header.plaintextSize
        for index in 0..<header.totalChunks {
            let plaintextLen = Int(min(Int64(chunkPlaintextBytes), remaining))
            let wireLen = nonceSize + plaintextLen + tagSize
            let wire = inHandle.readData(ofLength: wireLen)
            guard wire.count == wireLen else { throw Error.truncatedChunk }
            let nonce = wire.prefix(nonceSize)
            let tag = wire.suffix(tagSize)
            let ciphertext = wire.dropFirst(nonceSize).dropLast(tagSize)
            let plain = try openChunk(
                ciphertext: Data(ciphertext),
                tag: Data(tag),
                contentKey: contentKey,
                header: header,
                chunkIndex: UInt64(index),
                nonce: Data(nonce)
            )
            outHandle.write(plain)
            remaining -= Int64(plain.count)
        }
        guard remaining == 0 else { throw Error.chunkCountMismatch }
    }
}
